static_test_driver/autosequence.ino

```diff
@@ -1,171 +1,178 @@
-      // State that waits for cool-down to take place after a run is complete
-      digitalWrite(STATE_LED, (millis() % COOL_DOWN_LED_PERIOD) * 2 / COOL_DOWN_LED_PERIOD);
-      if (millis() - shutdown_time >= COOLDOWN_TIME) {
-        Serial.println(F("Run finished"));
-        SET_STATE(STAND_BY)
-        start_time = 0;
-      }
-      break;
-  }
-}
-
+// Timing
+#define PRESTAGE_PREP_TIME -500 // Time at which to open the prestage valves
+#define PRESTAGE_TIME      0
+#define MAINSTAGE_TIME     2000
+#define THRUST_CHECK_TIME  4000  // Time at which to start checking the engine is producing thrust, etc.
+
+#if CONFIGURATION == DEMO
+#define COUNTDOWN_DURATION 10000 // 10 sec
+#define RUN_TIME           10000 // 10 sec
+#define COOLDOWN_TIME      10000 // 10 sec
+#else
+#define COUNTDOWN_DURATION 60000 // 1 min
+#define RUN_TIME           12000 // 12 sec
+#define OX_LEADTIME        500 // 0.5 sec time for oxidizer to close before fuel
+#define COOLDOWN_TIME      60000 * 5 // 5 mins
+#endif
+
+// Limits
+// Max outlet temperature before triggering a shutdown
+#define MAX_COOLANT_TEMP 60
+
+// Min thrust that must be reached to avoid triggering a no-ignition shutdown
+#if CONFIGURATION == DEMO
+#define MIN_THRUST 50
+#elif CONFIGURATION == MK_1
+#define MIN_THRUST 2
+#elif CONFIGURATION == MK_2
+#define MIN_THRUST 1000
+#endif
+
+// State LED
+#define TERMINAL_COUNT_LED_PERIOD 250
+#define COOL_DOWN_LED_PERIOD 1000
+
+typedef enum {
+  STAND_BY,
+  TERMINAL_COUNT,
+  PRESTAGE_READY,
+  PRESTAGE,
+  MAINSTAGE,
+  COOL_DOWN
+} state_t;
+
+// Convenience
+#define SET_STATE(STATE) {\
+    state = STATE;          \
+    SEND(status, #STATE);   \
+  }
+
+long start_time = 0;
+long shutdown_time = 0;
+state_t state = STAND_BY;
+
+void start_countdown() {
+  if (!sensor_status) {
+    Serial.println(F("Countdown aborted due to sensor failure"));
+    SET_STATE(STAND_BY) // Set state to signal countdown was aborted
+  }
+  else if (valve_status[FUEL_PRE] ||
+           valve_status[FUEL_MAIN] ||
+           valve_status[OX_PRE] ||
+           valve_status[OX_MAIN]) {
+    Serial.println(F("Countdown aborted due to unexpected initial state"));
+    SET_STATE(STAND_BY) // Set state to signal countdown was aborted
+  }
+  else {
+    Serial.println(F("Countdown started"));
+    SET_STATE(TERMINAL_COUNT)
+    start_time = millis();
+  }
+}
+
+void abort_autosequence() {
+  Serial.println(F("Run aborted"));
+  switch (state) {
+    case STAND_BY:
+    case TERMINAL_COUNT:
+      SET_STATE(STAND_BY)
+      break;
+
+    case PRESTAGE_READY:
+      set_valve(FUEL_PRE, 0);
+      set_valve(OX_PRE, 0);
+      break;
+
+    case PRESTAGE:
+    case MAINSTAGE:
+      SET_STATE(COOL_DOWN)
+      set_valve(OX_PRE, 0);
+      set_valve(OX_MAIN, 0);
+      delay(200); //closes the oxygen valves slightly before the fuel valves like in a normal shutdown but with a delay function instead because who cares about 200ms of blocking code in an abort procedure?

```

Comment:
You should not have any sort of delay in the state machine - this will block data collection, etc. from occuring.

Reply:
I'll just get rid of the delay then, it's fine if they all close at the same time. I really don't have time at the moment to worry about the order they close in during an abort procedure.